pmlc/ast/ast_ops.cc

```diff
@@ -133,13 +134,47 @@ struct GatherOp : Intrinsic {
           "to be a constant float");
     }
 
-    TensorShape shape{tensor.elementType};
-    for (auto i = 0; i < axis.getValue(); i++) {
-      shape.sizes.push_back(tensor.sizes[i]);
+    auto mode = getIntegerValue(evaluator, operands[6]);
+    if (!mode) {
+      throw std::runtime_error("'gather' primitive expects the 'mode' argument "
+                               "to be a constant integer");
+    }
+
+    auto batch_dims = getIntegerValue(evaluator, operands[7]);
+    if (!batch_dims) {
+      throw std::runtime_error(
+          "'gather' primitive expects the 'batch_dims' argument "
+          "to be a constant integer");
     }
-    shape.sizes.insert(shape.sizes.end(), idxs.sizes.begin(), idxs.sizes.end());
-    for (auto i = axis.getValue() + 1; i < rank; i++) {
-      shape.sizes.push_back(tensor.sizes[i]);
+
+    TensorShape shape{tensor.elementType};
+    if (mode.getValue() == 0) {
+      for (auto i = 0; i < axis.getValue(); i++) {
+        shape.sizes.push_back(tensor.sizes[i]);
+      }
+      shape.sizes.insert(shape.sizes.end(), idxs.sizes.begin(),
+                         idxs.sizes.end());
+      for (auto i = axis.getValue() + 1; i < rank; i++) {
+        shape.sizes.push_back(tensor.sizes[i]);
+      }
+    } else {
+      /*
+      size_t firstDim = 1;
```

Comment:
Is the commented code will be used/updated in the future?

Reply:
This was intended for reshaping/multiplying the leading `batch_dims` into one dimension but as I mentioned above it will incur more complexity and have different behavior with `tf.gather_nd`, I temporarily comment it out. Will remove it later if we decide to adopt the approach of `tf.gather_nd`.